File: src/main/java/com/jcloisterzone/board/pointer/ScorePositionsFeaturePointer.java
package com.jcloisterzone.board.pointer;

import com.jcloisterzone.Immutable;
import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.BoardPointer;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.figure.Meeple;

import io.vavr.collection.Set;
import io.vavr.Tuple2;

/**
 * Points on feature on board or placed meeple.
 */
@Immutable
public class ScorePositionsFeaturePointer implements BoardPointer {

    private static final long serialVersionUID = 1L;

    private final FeaturePointer featurePointer;
    private final Set<Position> positions;

    public ScorePositionsFeaturePointer(FeaturePointer featurePointer, Set<Position> positions) {
        this.featurePointer = featurePointer;
        this.positions = positions;
    }

    public FeaturePointer asFeaturePointer() {
        return featurePointer;
    }

    public Position getPosition() {
        return featurePointer.getPosition();
    }

    public Set<Position> getPositions() {
        return positions;
    }

    @Override
    public String toString() {
        return "[x=" + featurePointer.getPosition().x + ",y=" + featurePointer.getPosition().y + "," + featurePointer.getFeature() + "," + featurePointer.getLocation() + "," + "positions=" + positions.mkString(",") + "]";
    }
}

File: src/main/ts/com/jcloisterzone/board/pointer/ScorePositionsFeaturePointer.ts
import type { Set } from "../../../../io/vavr/Set.js";
import type { Position } from "../Position.js";
import type { BoardPointer } from "./BoardPointer.js";
import type { FeaturePointer } from "./FeaturePointer.js";

/** A scoring source: a feature pointer plus the set of positions it spans. */
export class ScorePositionsFeaturePointer implements BoardPointer {
  constructor(
    private readonly featurePointer: FeaturePointer,
    private readonly positions: Set<Position>,
  ) {}

  asFeaturePointer(): FeaturePointer {
    return this.featurePointer;
  }

  getPosition(): Position {
    return this.featurePointer.getPosition();
  }

  getPositions(): Set<Position> {
    return this.positions;
  }

  toString(): string {
    const fp = this.featurePointer;
    return `[x=${fp.getPosition().x},y=${fp.getPosition().y},${fp.getFeature()},${fp.getLocation()},positions=${this.positions.mkString(",")}]`;
  }
}
